spec/job/adapters/sge/batch_spec.rb

```diff
@@ -0,0 +1,147 @@
+require "spec_helper"
+require "ood_core/job/adapters/sge"
+require "ood_core/job/adapters/sge/batch"
+
+def load_resource_file(file_name)
+  File.open(file_name, 'r') { |f| f.read }
+end
+
+describe OodCore::Job::Adapters::Sge::Batch do
+  subject(:batch) {described_class.new({:conf => '', :cluster => '', :bin => ''})}
+  let(:jobs_from_qstat) {[
+    OodCore::Job::Info.new( # Running job, w/ project
+      :id => '88',
+      :job_owner => 'vagrant',
+      :accounting_id => 'project_a',
+      :job_name => 'job_15',
+      :status => :running,
+      :procs => 1,
+      :queue_name => 'general.q',
+      :dispatch_time => DateTime.parse('2018-10-10T14:37:16').to_time.to_i,
+      :wallclock_limit => 360,
+      :wallclock_time => Time.now.to_i - DateTime.parse('2018-10-10T14:37:16').to_time.to_i
+    ),
+    OodCore::Job::Info.new( # Queued job, w/ project
+      :id => '1045',
+      :job_owner => 'vagrant',
+      :accounting_id => 'project_b',
+      :job_name => 'job_RQ',
+      :status => :queued,
+      :procs => 1,
+      :queue_name => 'general.q',
+      :submission_time => DateTime.parse('2018-10-09T18:47:05').to_time.to_i,
+      :wallclock_limit => 360,
+      :wallclock_time => 0
+    ),
+    OodCore::Job::Info.new( # Queued job w/o project
+      :id => '1046',
+      :job_owner => 'vagrant',
+      :job_name => 'job_RR',
+      :status => :queued,
+      :procs => 1,
+      :queue_name => 'general.q',
+      :submission_time => DateTime.parse('2018-10-09T18:47:05').to_time.to_i,
+      :wallclock_limit => 360,
+      :wallclock_time => 0
+    ),
+    OodCore::Job::Info.new( # Held job w/o project
+      :id => '44',
+      :job_owner => 'vagrant',
+      :job_name => 'c_d',
+      :status => :queued_held,
+      :procs => 1,
+      :queue_name => 'general.q',
+      :submission_time => DateTime.parse('2018-10-09T18:35:12').to_time.to_i,
+      :wallclock_limit => 360,
+      :wallclock_time => 0
+    )
+  ]}
+
+  let(:job_from_qacct) {
+    OodCore::Job::Info.new(
+      :accounting_id => nil,
+      :allocated_nodes => [],
+      :cpu_time => nil,
+      :dispatch_time => Time.parse('2018-10-25 13:16:29 +0000'),
+      :id => "1072",
+      :job_name => "job_7",
+      :job_owner => "vagrant",
+      :native => nil,
```

Comment:
In the other adapters we have returned the object representation of the original set of attributes returned from the adapter in this native hash. 

Since you aren't parsing the output of qstat using a parse-able column output, but rather XML, I can see why it might be preferable only to capture specific values. But it might be worth at least adding to native the sqe state.

The full native hash might look like this, given this XML:

```xml
    <job_list state="pending">
      <JB_job_number>1045</JB_job_number>
      <JAT_prio>0.25000</JAT_prio>
      <JB_name>job_RQ</JB_name>
      <JB_owner>vagrant</JB_owner>
      <JB_project>project_b</JB_project>
      <state>qw</state>
      <JB_submission_time>2018-10-09T18:47:05</JB_submission_time>
      <slots>1</slots>
      <full_job_name>job_RQ</full_job_name>
      <hard_request name="h_rt" resource_contribution="0.000000">360</hard_request>
      <hard_req_queue>general.q</hard_req_queue>
    </job_list> 
```

to

```ruby
{
  JB_job_number: 1045,
  JAT_prio: 0.25000,
  JB_name: "job_RQ",
  JB_owner: "vagrant",
  JB_project: "project_b",
  state: "qw",
  JB_submission_time: "2018-10-09T18:47:05",
  slots: 1,
  full_job_name: "job_RQ",
  hard_request: 360,
  hard_request_name: "h_rt",
  hard_request_resource_contribution: "0.000000",
  hard_req_queue: "general.q"
}
```

And we would add an `extended_data_sge` method to Active Jobs that contained this SGE specific information. Here is the Slurm example: https://github.com/OSC/ood-activejobs/blob/b77a6889b61fe03af13902bcb0c0cccada2530da/app/models/jobstatusdata.rb#L101-L128/jobstatusdata.rb#L101-L128

And that is what appears in the table when you expand the details of a job in the Active Jobs app.

Note: I think it is worth stating that sending all this extra information in the native hash costs something, and the default information for Job::Info might be good enough for many sites, or only a subset of the native hash might be good enough, and thus for these reasons it might be preferable to have a mode where adapters do not set the native hash on the Job::Info object. That is one consideration for the performance improvements after this OOD release.

Reply:
Also, removing adapter specific knowledge from ActiveJobs, which I think can be done.

Reply:
Two benefits of a native hash:

1. Users can see adapter specific information that may not have a corresponding attr in the Info object
2. Sometimes we parse a value into a generic format and that parsing is incorrect or strips valuable information. The original value can be accessed by the user if that value is in the native hash

Reply:
While we don't do it in other adapters yet, we could omit from the native hash any values that are identical duplicates on job info attrs. That would save a little memory/time.

Reply:
Or maybe for the sake of time, just return {} for the native hash, write up an issue to add extended data to SGE, we can then add this in 1.5 release after we do performance improvements.

Reply:
Opened #114.

Reply:
Updated `#info` and `#info_all` to always return an empty hash for the time being.